File: crates/netcfgd-apply/src/dhcpcd_control.rs
//! Asking a running dhcpcd which configuration file it was started with.
//!
//! **The one backend whose mark cannot be read from the process.** netcfgd
//! recovers its supplicant, and now its udhcpc, by finding a path it chose as a
//! whole `argv` element (decision 0140). dhcpcd calls `setproctitle` and
//! destroys both: measured, `/proc/<pid>/cmdline` reads `dhcpcd: wlp0s20f3
//! [ip4]`, and the environment block comes back 4494 bytes of NUL against a
//! control process that kept its variables. Nothing netcfgd passed survives in
//! the process image.
//!
//! What does survive is dhcpcd's own memory of its `-f` argument, which it
//! recites verbatim -- symlink and all, with no `realpath` -- to anyone who
//! asks `--getconfigfile` on its control socket. So netcfgd starts dhcpcd with
//! a `-f` under its own run directory and asks for it back. Decision 0143.
//!
//! **Three things measured that the obvious implementation gets wrong**, each
//! of which breaks a platform this exists to serve:
//!
//! - **The privileged socket, not the unprivileged one.** dhcpcd 10.5.0
//!   removed `<iface>-4.unpriv.sock` outright -- "a breaking ABI change" in its
//!   own commit message -- and Debian sid ships 10.5.2 today. netcfgd is root,
//!   so the privileged socket is available on every version that has a socket
//!   at all, and answers this command identically.
//! - **The length prefix is a native `size_t`.** dhcpcd's `control.c` writes
//!   `iov[0].iov_len = sizeof(size_t)`: eight bytes on amd64, **four** on
//!   32-bit ARM, and big-endian on a big-endian MIPS. Parsing it as a `u64`
//!   little-endian works on the developer's machine and on nothing else this
//!   targets, so the reply is read to its first NUL under a byte cap instead.
//! - **An unknown command does not fail, it hangs.** `--getinterfaces`,
//!   `--isprivileged` and a bare `-q` were each measured to produce no reply
//!   *and no close*, past a four-second wait, on both sockets. A probe without
//!   a deadline is a daemon that stops reconciling.

use std::io::{Read, Write};
use std::os::unix::net::UnixStream;
use std::path::PathBuf;
use std::time::Duration;

/// How long to wait for a reply.
///
/// Generous rather than tuned. The command is answered by dhcpcd's separate
/// `[control proxy]` process, which replied in 0.00s even with the main dhcpcd
/// stopped with `SIGSTOP` -- only stopping *every* dhcpcd process silenced it.
/// So the deadline exists for the wedged case, not the ordinary one, and
/// waiting longer buys nothing but a slower refusal.
const DEADLINE: Duration = Duration::from_millis(250);

/// The longest reply worth reading.
///
/// A path, so `PATH_MAX` and a little. This is the cap that stands in for the
/// length prefix netcfgd deliberately does not parse.
const REPLY_MAX: usize = 4200;

/// Where dhcpcd puts its control socket for one interface and family.
///
/// Privileged first. The unprivileged one is a fallback for nothing in
/// particular -- netcfgd is root -- and is tried only because a machine may
/// have tightened the privileged socket's mode.
fn socket_paths(run_dir: &str, iface: &str, family: &str) -> Vec<PathBuf> {
	let base = PathBuf::from(run_dir);
	vec![
		base.join(format!("{iface}-{family}.sock")),
		base.join(format!("{iface}-{family}.unpriv.sock")),
	]
}

/// The configuration file a running dhcpcd was started with.
///
/// `None` covers every way of not knowing: no socket, nothing listening, a
/// version that does not answer, a reply that does not parse, or the deadline.
/// **`None` is not "somebody else's"** -- it is "netcfgd could not tell", and
/// the caller must treat the two differently, which is 0074's rule and 0141's
/// default.
#[must_use]
pub fn config_file_of(run_dir: &str, iface: &str, family: &str) -> Option<String> {
	for path in socket_paths(run_dir, iface, family) {
		let Ok(mut stream) = UnixStream::connect(&path) else {
			continue;
		};
		if stream.set_read_timeout(Some(DEADLINE)).is_err()
			|| stream.set_write_timeout(Some(DEADLINE)).is_err()
		{
			continue;
		}
		// Arguments are NUL-separated with a newline before the final NUL.
		if stream.write_all(b"--getconfigfile\n\0").is_err() {
			continue;
		}
		// Read what arrives within the deadline, then parse. Not "read to the
		// first NUL": the length prefix's low byte is printable for any
		// ordinary path -- measured, a 33-byte path gives `22 00 00 00 00 00
		// 00 00` -- so a scan stopping at the first NUL stops after one
		// character. dhcpcd does not close the connection either, so this
		// reads once under the timeout rather than to EOF.
		let mut buffer = vec![0_u8; REPLY_MAX];
		let read = match stream.read(&mut buffer) {
			Ok(0) | Err(_) => continue,
			Ok(count) => count,
		};
		buffer.truncate(read);
		let Some(reply) = payload(&buffer) else {
			continue;
		};
		if !reply.is_empty() {
			return Some(reply);
		}
	}
	None
}

/// The path out of one control-socket reply.
///
/// The frame is a native-width length then a NUL-terminated string, and netcfgd
/// parses only the second half. The width is the platform's -- eight bytes on
/// amd64, **four** on 32-bit ARM, big-endian on a big-endian MIPS -- so a
/// reader that decodes it works on the machine it was written on and on none of
/// the others this project targets.
///
/// Instead: the last run of printable bytes. A filesystem path holds no NUL and
/// no control character, dhcpcd sends exactly one string, and the prefix is
/// binary -- so the tail is the answer whatever width the prefix had.
fn payload(bytes: &[u8]) -> Option<String> {
	let end = bytes.iter().rposition(|byte| *byte >= 0x20)? + 1;
	let start = bytes[..end]
		.iter()
		.rposition(|byte| *byte < 0x20)
		.map_or(0, |at| at + 1);
	let text = String::from_utf8_lossy(&bytes[start..end])
		.trim()
		.to_owned();
	(!text.is_empty()).then_some(text)
}

#[cfg(test)]
mod tests {
	use super::payload;

	/// The exact bytes dhcpcd 10.1.0 sent, measured in a namespace.
	#[test]
	fn the_measured_reply_parses() {
		let mut wire = vec![0x22, 0, 0, 0, 0, 0, 0, 0];
		wire.extend_from_slice(b"/run/netcfgd/dhcpcd/probe0-4.conf\0");
		assert_eq!(
			payload(&wire).as_deref(),
			Some("/run/netcfgd/dhcpcd/probe0-4.conf")
		);
	}

	/// **The bug this replaced.** Reading to the first NUL stops after the
	/// length prefix's low byte, which is printable for any ordinary path.
	#[test]
	fn the_printable_length_byte_is_not_the_answer() {
		let mut wire = vec![0x11, 0, 0, 0, 0, 0, 0, 0];
		wire.extend_from_slice(b"/etc/dhcpcd.conf\0");
		assert_eq!(payload(&wire).as_deref(), Some("/etc/dhcpcd.conf"));
	}

	/// A four-byte prefix, which is what a 32-bit platform sends.
	#[test]
	fn a_narrower_prefix_parses_the_same() {
		let mut wire = vec![0x11, 0, 0, 0];
		wire.extend_from_slice(b"/etc/dhcpcd.conf\0");
		assert_eq!(payload(&wire).as_deref(), Some("/etc/dhcpcd.conf"));
	}

	#[test]
	fn nothing_printable_is_no_answer() {
		assert_eq!(payload(&[0, 0, 0, 0]), None);
		assert_eq!(payload(&[]), None);
	}
}

File: c/src/apply/dhcpcd_control.c
/*
 * dhcpcd_control.c -- asking a running dhcpcd which configuration file it was
 * started with.
 *
 * The reasoning is in `apply.h`, where a caller will read it: dhcpcd calls
 * `setproctitle` and destroys both the command line and the environment
 * netcfgd started it with, so the only surviving mark is dhcpcd's own memory
 * of its `-f` argument, which it recites on its control socket.
 *
 * WHAT THIS OPENS, AND WHAT STOPS IT
 *   One `AF_UNIX` stream socket at a time, under a run directory the caller
 *   names, with a send and a receive deadline set before a byte is written.
 *   It spawns nothing, signals nothing and loops only while a read is still
 *   making progress into a buffer with a ceiling -- an unknown command on this
 *   socket produces no reply *and no close*, which is what the deadline is
 *   for.
 */
#include "ncfg/apply.h"

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <sys/un.h>
#include <unistd.h>

/* Arguments are NUL-separated with a newline before the final NUL, which is
 * the framing dhcpcd's `control.c` reads. */
static const char get_config_file[] = "--getconfigfile\n\0";

/* The bytes of it that go on the wire: everything but the terminator C adds
 * after the one dhcpcd's framing asks for. */
#define GET_CONFIG_FILE_BYTES (sizeof(get_config_file) - 1u)

/* ------------------------------------------------------------------------ *
 * The reply
 * ------------------------------------------------------------------------ */

int ncfg_dhcpcd_control_payload(const void *bytes, size_t length, char *out, size_t out_size)
{
	const unsigned char *at = bytes;
	size_t               end;
	size_t               start;
	size_t               run;

	if (!out || out_size == 0u) {
		return 0;
	}
	out[0] = '\0';
	if (!at || length == 0u) {
		return 0;
	}
	/* The last run of printable bytes. A filesystem path holds no NUL and no
	 * control character, dhcpcd sends exactly one string, and the length
	 * prefix in front of it is binary -- so the tail is the answer whatever
	 * width that prefix had. */
	end = length;
	while (end > 0u && at[end - 1u] < 0x20u) {
		end--;
	}
	if (end == 0u) {
		return 0;
	}
	/*
	 * **A run that reaches the end of what was read is not an answer.** dhcpcd
	 * terminates the string, so a reply with no terminator in it is a reply
	 * that has not all arrived -- and believing it is how the length prefix's
	 * own low byte came to be reported as a path.
	 */
	if (end == length) {
		return 0;
	}
	start = end;
	while (start > 0u && at[start - 1u] >= 0x20u) {
		start--;
	}
	run = end - start;
	while (run > 0u && at[start] == ' ') {
		start++;
		run--;
	}
	while (run > 0u && at[start + run - 1u] == ' ') {
		run--;
	}
	if (run == 0u || run + 1u > out_size) {
		return 0;
	}
	/*
	 * **Absolute, or it is not an answer.** The one thing every caller does
	 * with this is compare it against a path netcfgd composed, and a reply
	 * that is not a path at all compares unequal -- which reads as "somebody
	 * else's dhcpcd" rather than as "netcfgd could not tell". Those two have
	 * different consequences (0141), so the difference is decided here.
	 */
	if (at[start] != (unsigned char)'/') {
		return 0;
	}
	memcpy(out, at + start, run);
	out[run] = '\0';
	return 1;
}

/* ------------------------------------------------------------------------ *
 * The socket
 * ------------------------------------------------------------------------ */

/* A copy of `text`, or NULL. */
static char *duplicate(const char *text)
{
	size_t length = strlen(text) + 1u;
	char  *copy = malloc(length);

	if (copy) {
		memcpy(copy, text, length);
	}
	return copy;
}

/* Ask one socket, and answer what it recited. NULL for every way of not
 * knowing. */
static char *ask(const char *path)
{
	struct sockaddr_un  where;
	struct timeval      deadline;
	unsigned char       reply[NCFG_DHCPCD_REPLY_MAX];
	char                answer[NCFG_DHCPCD_REPLY_MAX];
	size_t              filled = 0u;
	size_t              sent = 0u;
	int                 fd;

	if (strlen(path) >= sizeof(where.sun_path)) {
		return NULL;
	}
	fd = socket(AF_UNIX, SOCK_STREAM, 0);
	if (fd < 0) {
		return NULL;
	}
	memset(&where, 0, sizeof(where));
	where.sun_family = AF_UNIX;
	(void)snprintf(where.sun_path, sizeof(where.sun_path), "%s", path);

	memset(&deadline, 0, sizeof(deadline));
	deadline.tv_usec = NCFG_DHCPCD_DEADLINE_MILLISECONDS * 1000;
	if (setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &deadline, (socklen_t)sizeof(deadline)) != 0 ||
	    setsockopt(fd, SOL_SOCKET, SO_SNDTIMEO, &deadline, (socklen_t)sizeof(deadline)) != 0) {
		(void)close(fd);
		return NULL;
	}
	if (connect(fd, (const struct sockaddr *)&where, (socklen_t)sizeof(where)) != 0) {
		(void)close(fd);
		return NULL;
	}
	while (sent < GET_CONFIG_FILE_BYTES) {
		ssize_t put = write(fd, get_config_file + sent, GET_CONFIG_FILE_BYTES - sent);

		if (put < 0 && errno == EINTR) {
			continue;
		}
		if (put <= 0) {
			(void)close(fd);
			return NULL;
		}
		sent += (size_t)put;
	}

	/*
	 * Read until the reply holds a whole answer, or until the deadline.
	 *
	 * **Not "read once", which is what the Rust does.** A reply split between
	 * its length prefix and its string leaves the prefix's printable low byte
	 * as the last run of printable bytes, and that is an answer the parser
	 * above refuses -- so this asks for the rest rather than returning
	 * something that is not a path. dhcpcd does not close the connection
	 * either, so an answer that never completes ends at the receive timeout
	 * rather than at end of file.
	 */
	while (filled < sizeof(reply)) {
		ssize_t got;

		if (ncfg_dhcpcd_control_payload(reply, filled, answer, sizeof(answer))) {
			(void)close(fd);
			return duplicate(answer);
		}
		got = read(fd, reply + filled, sizeof(reply) - filled);
		if (got < 0 && errno == EINTR) {
			continue;
		}
		if (got <= 0) {
			break;
		}
		filled += (size_t)got;
	}
	if (ncfg_dhcpcd_control_payload(reply, filled, answer, sizeof(answer))) {
		(void)close(fd);
		return duplicate(answer);
	}
	(void)close(fd);
	return NULL;
}

char *ncfg_dhcpcd_config_file_of(const char *run_dir, const char *interface, const char *family)
{
	/*
	 * Privileged first. The unprivileged one is a fallback for nothing in
	 * particular -- netcfgd is root -- and is tried only because a machine may
	 * have tightened the privileged socket's mode. dhcpcd 10.5.0 removed it
	 * outright, so on a current Debian the second path is never there.
	 */
	static const char *const suffixes[] = { "sock", "unpriv.sock" };
	size_t                   at;

	if (!run_dir || !interface || !family) {
		return NULL;
	}
	for (at = 0u; at < sizeof(suffixes) / sizeof(*suffixes); at++) {
		char path[512];
		char *found;
		int   written;

		written = snprintf(path, sizeof(path), "%s/%s-%s.%s", run_dir, interface, family,
		    suffixes[at]);
		if (written < 0 || (size_t)written >= sizeof(path)) {
			continue;
		}
		found = ask(path);
		if (found) {
			return found;
		}
	}
	return NULL;
}
